File: events/domain/entities/event.go
package entities

import (
	"events/domain"
	domain_errors "events/domain/errors"
	"time"

	uuid "github.com/satori/go.uuid"
)

type EventStatus string

const (
	StatusAvailable EventStatus = "available"
	StatusSoldOut   EventStatus = "sold_out"
	StatusFinished  EventStatus = "finished"
	StatusCanceled  EventStatus = "canceled"
)

type EventVisibility string

const (
	VisibilityPrivate     EventVisibility = "private"
	VisibilityPublic      EventVisibility = "public"
	VisibilityInvitedOnly EventVisibility = "invited_only"
)

type Event struct {
	domain.Entity
	Name               string          `json:"name,omitempty"`
	Description        string          `json:"description,omitempty"`
	CoverUrl           string          `json:"cover_url,omitempty"`
	OrganizerAccountId string          `json:"organizer_account_id,omitempty"`
	AgeGroup           int             `json:"age_group,omitempty"`
	Status             EventStatus     `json:"status,omitempty"`
	Location           *EventLocation  `json:"location" gorm:"foreignKey:EventId;references:Id"`
	Tickets            []Ticket        `json:"ticket_options,omitempty" gorm:"foreignKey:EventId;references:Id"`
	StartDateTime      time.Time       `json:"start_datetime,omitempty"`
	EndDateTime        time.Time       `json:"end_datetime,omitempty"`
	CategoryId         string          `json:"category,omitempty" gorm:"foreignKey:CategoryId"`
	SubjectId          string          `json:"subject,omitempty" gorm:"foreignKey:SubjectId"`
	Visibility         EventVisibility `json:"visibility,omitempty"`
}

func NewEvent(
	name, description, coverUrl, organizerAccountId string,
	ageGroup int, location *EventLocation,
	tickets []Ticket, startDateTime, endDateTime string, categoryId,
	subjectId, visibility string,
) (*Event, domain_errors.DomainErr) {
	var err error

	event := new(Event)

	event.Id = uuid.NewV4().String()
	event.CreatedAt = time.Now()
	event.UpdatedAt = time.Now()

	event.Name = name
	event.Description = description
	event.CoverUrl = coverUrl
	event.OrganizerAccountId = organizerAccountId
	event.AgeGroup = ageGroup
	event.Location = location
	event.Tickets = tickets
	event.CategoryId = categoryId
	event.SubjectId = subjectId

	event.StartDateTime, err = time.Parse(time.RFC3339, startDateTime)
	if err != nil {
		return nil, domain_errors.NewValidationError(
			`Event's dates must be in "RFC3339" format`,
			"StartDateTime", startDateTime)
	}

	event.EndDateTime, err = time.Parse(time.RFC3339, endDateTime)
	if err != nil {
		return nil, domain_errors.NewValidationError(
			`Event's dates must be in "RFC3339" format`,
			"EndDateTime", endDateTime)
	}

	switch visibility {
	case string(VisibilityPrivate), string(VisibilityPublic), string(VisibilityInvitedOnly):
		event.Visibility = EventVisibility(visibility)
	default:
		{
			return nil, domain_errors.NewValidationError(
				"Events' visibility must be: private, public or invited_only",
				"Visibility", visibility)
		}
	}

	if err = event.validate(); err != nil {
		return nil, err
	}

	return event, nil
}

func (e *Event) validate() domain_errors.DomainErr {
	if len(e.Name) == 4 || len(e.Name) > 255 {
		return domain_errors.NewValidationError(
			"Events name must have at least of 4 characters and at most of 255",
			"name",
			e.Name)
	}

	return nil
}

File: _events/src/modules/events/domain/event.ts
import { Entity } from '@/shared/entity'

export type EventStatus = 'available' | 'finished' | 'sold_out' | 'canceled'

export type EventVisibility = 'public' | 'private' | 'invited_only'

export interface EventData {
  name: string
  description: string
  coverUrl: string
  organizerAccountId: string
  ageGroup: number
  status: EventStatus
  locationId: string
  startDateTime: Date
  endDateTime: Date
  categoryId: string
  subjectId: string
  visibility: EventVisibility
}

export class Event extends Entity<EventData> {
  constructor(data: EventData, id: string) {
    super(data, id)
  }

  public get name(): string {
    return this.data.name
  }

  public get description(): string {
    return this.data.description
  }

  public get coverUrl(): string {
    return this.data.coverUrl
  }

  public get organizerAccountId(): string {
    return this.data.organizerAccountId
  }

  public get agrGroup(): number {
    return this.data.ageGroup
  }

  public get status(): string {
    return this.data.status
  }

  public get locationId(): string {
    return this.data.locationId
  }

  public get startDateTime(): Date {
    return this.data.startDateTime
  }

  public get endDateTime(): Date {
    return this.data.endDateTime
  }

  public get categoryId(): string {
    return this.data.categoryId
  }

  public get subjectId(): string {
    return this.data.subjectId
  }

  public get visibility(): string {
    return this.data.visibility
  }
}
